File: app/src/main/java/com/example/kirppis/MainActivity.java
package com.example.kirppis;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import com.firebase.ui.auth.AuthUI;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private DatabaseReference db;
    private final ArrayList<CustomView> itemList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        FirebaseAuth auth = FirebaseAuth.getInstance();
        db = FirebaseDatabase.getInstance().getReference();

        //kirjautuminen
        FirebaseUser currentUser = auth.getCurrentUser();
        if (currentUser == null) {
            Log.d("debuggi", "not signed in");
            Intent intent = new Intent(this, SignIn.class);
            startActivity(intent);
        }
        else{
            Log.d("debuggi", "signed in " + currentUser);
            loadItems();
        }
    }

    //tuotelistan lataus firebasesta
    private void loadItems() {
        db.child("items").addListenerForSingleValueEvent((new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                Log.d("debuggi" ,"item count: " + snapshot.getChildrenCount());

                //ladataan ensin databasesta tuotteet itemListiin
                for (DataSnapshot postSnapshot: snapshot.getChildren()) {
                    itemList.add(new CustomView(postSnapshot.getValue(Item.class).getImage(),
                                                postSnapshot.getValue(Item.class).getName(),
                                      postSnapshot.getValue(Item.class).getPrice() + "€"));
                }
                // Now create the instance of the CustomArrayAdapter and pass
                // the context and arrayList created above
                CustomArrayAdapter customArrayAdapter = new CustomArrayAdapter(MainActivity.this, itemList);

                // create the instance of the ListView to set the CustomArrayAdapter
                ListView listView = findViewById(R.id.listView);

                // set the CustomArrayAdapter for ListView
                listView.setAdapter(customArrayAdapter);

                //viedään tuotteen id ItemListingiin klikattaessa sitä
                listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                        goToItemListing(i);
                    }
                });
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Log.e("The read failed: " ,error.toString());
            }
        }));
    }

    public void goToShoppingCart(View view){
        Intent intent = new Intent(this, ShoppingCart.class);
        startActivity(intent);
    }

    public void goToItemListing(int i){
        Intent intent = new Intent(this, ItemListing.class);
        intent.putExtra("item id", String.valueOf(i));
        startActivity(intent);
    }

    public void signOut(View view) {
        AuthUI.getInstance().signOut(this);
        Intent intent = new Intent(this, SignIn.class);
        startActivity(intent);
    }
}

File: app/src/main/java/com/example/kirppis/MainActivity.kt
package com.example.kirppis

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView.OnItemClickListener
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import com.firebase.ui.auth.AuthUI
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*

class MainActivity : AppCompatActivity() {
    private var db: DatabaseReference? = null
    private val itemList = ArrayList<CustomView>()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val auth = FirebaseAuth.getInstance()
        db = FirebaseDatabase.getInstance().reference

        //kirjautuminen
        val currentUser = auth.currentUser
        if (currentUser == null) {
            Log.d("debuggi", "not signed in")
            val intent = Intent(this, SignIn::class.java)
            startActivity(intent)
        } else {
            Log.d("debuggi", "signed in $currentUser")
            loadItems()
        }
    }

    //tuotelistan lataus firebasesta
    private fun loadItems() {
        db!!.child("items").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                Log.d("debuggi", "item count: " + snapshot.childrenCount)

                //ladataan ensin databasesta tuotteet itemListiin
                for (postSnapshot in snapshot.children) {
                    itemList.add(
                        CustomView(
                            postSnapshot.getValue(Item::class.java)!!.image,
                            postSnapshot.getValue(Item::class.java)!!.name,
                            postSnapshot.getValue(Item::class.java)!!.price + "€"
                        )
                    )
                }
                // Now create the instance of the CustomArrayAdapter and pass
                // the context and arrayList created above
                val customArrayAdapter = CustomArrayAdapter(this@MainActivity, itemList)

                // create the instance of the ListView to set the CustomArrayAdapter
                val listView = findViewById<ListView>(R.id.listView)

                // set the CustomArrayAdapter for ListView
                listView.adapter = customArrayAdapter

                //viedään tuotteen id ItemListingiin klikattaessa sitä
                listView.onItemClickListener =
                    OnItemClickListener { adapterView, view, i, l -> goToItemListing(i) }
            }

            override fun onCancelled(error: DatabaseError) {
                Log.e("The read failed: ", error.toString())
            }
        })
    }

    fun goToShoppingCart(view: View?) {
        val intent = Intent(this, ShoppingCart::class.java)
        startActivity(intent)
    }

    fun goToItemListing(i: Int) {
        val intent = Intent(this, ItemListing::class.java)
        intent.putExtra("item id", i.toString())
        startActivity(intent)
    }

    fun signOut(view: View?) {
        AuthUI.getInstance().signOut(this)
        val intent = Intent(this, SignIn::class.java)
        startActivity(intent)
    }
}